mycroft/util/lang/parse_en.py

```diff
@@ -14,235 +14,515 @@
-    17: 'seventeenth',
-    18: 'eighteenth',
-    19: 'nineteenth',
-    20: 'twentieth',
-    30: 'thirtieth',
-    40: "fortieth",
-    50: "fiftieth",
-    60: "sixtieth",
-    70: "seventieth",
-    80: "eightieth",
-    90: "ninetieth",
-    10e3: "hundredth",
-    1e3: "thousandth",
-    1e6: "millionth",
-    1e12: "billionth",
-    1e18: "trillionth",
-    1e24: "quadrillionth",
-    1e30: "quintillionth",
-    1e36: "sextillionth",
-    1e42: "septillionth",
-    1e48: "octillionth",
-    1e54: "nonillionth",
-    1e60: "decillionth"
-    # TODO > 1e60
-}
+from mycroft.util.lang.parse_common import is_numeric, look_for_fractions
+from mycroft.util.lang.common_data_en import _ARTICLES, _NUM_STRING_EN, \
+    _LONG_ORDINAL_STRING_EN, _LONG_SCALE_EN, \
+    _SHORT_SCALE_EN, _SHORT_ORDINAL_STRING_EN
 
+import re
 
-def extractnumber_en(text, short_scale=True, ordinals=False):
+
+def _invert_dict(original):
     """
-    This function extracts a number from a text string,
-    handles pronunciations in long scale and short scale
+    Produce a dictionary with the keys and values
+    inverted, relative to the dict passed in.
 
-    https://en.wikipedia.org/wiki/Names_of_large_numbers
+    Args:
+        original dict: The dict like object to invert
+
+    Returns:
+        dict
+
+    """
+    return {value: key for key, value in original.items()}
+
+
+def _generate_plurals(originals):
+    """
+    Return a new set or dict containing the original values,
+    all with 's' appended to them.
 
     Args:
-        text (str): the string to normalize
+        originals set(str) or dict(str, any): values to pluralize
+
+    Returns:
+        set(str) or dict(str, any)
+
+    """
+    if isinstance(originals, dict):
+        return {key + 's': value for key, value in originals.items()}
+    return {value + "s" for value in originals}
+
+
+# negate next number (-2 = 0 - 2)
+_NEGATIVES = {"negative", "minus"}
+
+# sum the next number (twenty two = 20 + 2)
+_SUMS = {'twenty', '20', 'thirty', '30', 'forty', '40', 'fifty', '50',
+         'sixty', '60', 'seventy', '70', 'eighty', '80', 'ninety', '90'}
+
+_MULTIPLIES_LONG_SCALE_EN = set(_LONG_SCALE_EN.values()) | \
+                            _generate_plurals(_LONG_SCALE_EN.values())
+
+_MULTIPLIES_SHORT_SCALE_EN = set(_SHORT_SCALE_EN.values()) | \
+                             _generate_plurals(_SHORT_SCALE_EN.values())
+
+
+# split sentence parse separately and sum ( 2 and a half = 2 + 0.5 )
+_FRACTION_MARKER = {"and"}
+
+# decimal marker ( 1 point 5 = 1 + 0.5)
+_DECIMAL_MARKER = {"point", "dot"}
+
+_STRING_NUM_EN = _invert_dict(_NUM_STRING_EN)
+_STRING_NUM_EN.update(_generate_plurals(_STRING_NUM_EN))
+_STRING_NUM_EN.update({
+    "half": 0.5,
+    "halves": 0.5,
+    "couple": 2
+})
+
+_STRING_SHORT_ORDINAL_EN = _invert_dict(_SHORT_ORDINAL_STRING_EN)
+_STRING_LONG_ORDINAL_EN = _invert_dict(_LONG_ORDINAL_STRING_EN)
+
+
+class _Token():
+    """
+    Map a word to an index in a string.
```

Comment:
index in a string makes it seem to be a string index instead of word index as it seems to be used. As the index is basically user defined Maybe just cut the _in a string_ part.

A note, if you don't expect to add extra information to the objects a namedtuple would work pretty similarly to this.